[SelectionDAG] Add `STRICT_BF16_TO_FP` and `STRICT_FP_TO_BF16`

shiltian opened the pull request:
This patch adds the support for `STRICT_BF16_TO_FP` and `STRICT_FP_TO_BF16`.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3248,12 +3249,18 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
+  case ISD::STRICT_BF16_TO_FP:
+    // When strict mode is enforced we can't do expansion because it
+    // does not honor the "strict" properties.
+    if (TLI.isStrictFPEnabled())
```
I don't understand why we have isStrictFPEnabled, but I think this check (and the hook more broadly) should be deleted. It's like we have 3 parallel mechanisms for supporting strictfp and this one makes the least sense. 

As far as this check goes, the target support of strictfp doesn't matter. This is going to fail to raise the required exceptions either way.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3271,8 +3278,14 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
-    SDValue Op = Node->getOperand(0);
+    SDValue Op = Node->getOperand(Node->getOpcode() == ISD::FP_TO_BF16 ? 0 : 1);
     if (Op.getValueType() != MVT::f32)
       Op = DAG.getNode(ISD::FP_ROUND, dl, MVT::f32, Op,
```
this needs to use STRICT_FP_ROUND in the strict case

phoebewang commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3248,12 +3249,18 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
       Results.push_back(Tmp1);
     break;
   }
+  case ISD::STRICT_BF16_TO_FP:
```
We cannot support `STRICT_BF16_TO_FP` by `SHL` because it cannot generate expected exceptions.

shiltian replied:
This is more like a "fallback" to the non-strict variant when the strict FP is not enforced, which is the purpose of the check below. We have something similar in this big `switch`, such as `case ISD::STRICT_FP_EXTEND`.

phoebewang replied:
They are different. The comment says the expansion is not allowed when strict mode is enforced. Here you just do the expansion unconditionally.

shiltian replied:
Yeah, I did have a check earlier but removed it since I think https://github.com/llvm/llvm-project/pull/80056#discussion_r1472611861 makes sense.

arsenm commented on llvm/test/CodeGen/AMDGPU/llvm.is.fpclass.bf16.ll:
```diff
@@ -1094,11 +1094,52 @@ define <4 x i1> @isnan_v4bf16(<4 x bfloat> %x) nounwind {
+ ; GFX11CHECK-NEXT:    v_cmp_lt_i16_e32 vcc_lo, 0x7f80, v0
+ ; GFX11CHECK-NEXT:    v_cndmask_b32_e64 v0, 0, 1, vcc_lo
+ ; GFX11CHECK-NEXT:    s_setpc_b64 s[30:31]
+   %1 = call i1 @llvm.is.fpclass.bf16(bfloat %x, i32 3) strictfp ; nan
```
don't need the strictfp callsite attribute

arsenm reviewed (commented):
To make progress, can you split the non-lib call expansion to a separate change? It's easier to start with the boilerplate and case that doesn't require writing new infrastructure

shiltian commented:
> To make progress, can you split the non-lib call expansion to a separate change? It's easier to start with the boilerplate and case that doesn't require writing new infrastructure

Removed the non-lib call expansion for now. However, I don't know how to test it now because there is no feasible lowering that I can test.

shiltian commented:
@phoebewang Does X86 have lib calls for `STRICT_BF16_TO_FP` and `STRICT_FP_TO_BF16`?

phoebewang commented:
> @phoebewang Does X86 have lib calls for `STRICT_BF16_TO_FP` and `STRICT_FP_TO_BF16`? I'm trying to find a target to test the lowering via the lib call path.

Do you mean this? https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/RuntimeLibcalls.def#L322

I see you use the same library as none strict one. But we don't provide libcall for extend.

shiltian commented:
I'm not sure how to design a valid lowering that can test the lib call path. Designed a similar X86 test case as the one from `llvm/test/CodeGen/X86/half-constrained.ll`, as shown below:
```
define void @float_to_bfloat(float %0) strictfp {
  %2 = tail call bfloat @llvm.experimental.constrained.fptrunc.bfloat.f32(float %0, metadata !"round.tonearest", metadata !"fpexcept.strict") #0
  store bfloat %2, ptr @a, align 2
  ret void
}
```
If the type is `half`, it will go through the lib call path. However, if it is `bfloat`, it will not (I believe it goes through soft promotion).

phoebewang commented:
> I'm not sure how to design a valid lowering that can test the lib call path. Designed a similar X86 test case as the one from `llvm/test/CodeGen/X86/half-constrained.ll`, as shown below:
> 
> ```
> define void @float_to_bfloat(float %0) strictfp {
>   %2 = tail call bfloat @llvm.experimental.constrained.fptrunc.bfloat.f32(float %0, metadata !"round.tonearest", metadata !"fpexcept.strict") #0
>   store bfloat %2, ptr @a, align 2
>   ret void
> }
> ```
> 
> If the type is `half`, it will go through the lib call path. However, if it is `bfloat`, it will not (I believe it goes through soft promotion).

The example can pass if you add expand to X86ISelLowering.cpp
```
setOperationAction(ISD::STRICT_FP_TO_BF16, MVT::f32, Expand);
```

shiltian commented on llvm/test/CodeGen/X86/bfloat-constrained.ll:
```diff
@@ -0,0 +1,98 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
This file was adapted from its `half` brother. I removed the `f16c` attribute tests because I guess they will not affect `bfloat`. I commented out those `fpext` tests since we can't support it right now.

phoebewang replied:
You need to not make it `Custom` for `Subtarget.hasF16C()`

arsenm commented on llvm/test/CodeGen/X86/bfloat-constrained.ll:
```diff
@@ -0,0 +1,98 @@
+; declare float @llvm.experimental.constrained.fadd.f32(float, float, metadata, metadata)
+declare bfloat @llvm.experimental.constrained.fptrunc.bfloat.f32(float, metadata, metadata)
+declare bfloat @llvm.experimental.constrained.fptrunc.bfloat.f64(double, metadata, metadata)
+
```
Test vectors too?

phoebewang replied:
Test STRICT_FP16_TO_FP? I think the key is we have to lower it to a new libcall instead of expand it through integer operation.
I'm also interested in test with `-mattr=+avx512bf16,+avx512vl`, because the VCVTNEPS2BF16 doesn't raise exception. We should lower it to libcall too.

shiltian replied:
`STRICT_FP16_TO_FP` would be out of the scope of this patch (as well as the file) since it already exists.

phoebewang replied:
Sorry, I meat to `STRICT_BF16_TO_FP`

shiltian replied:
Is there any existing GNU lib functions for that? It looks like from https://gcc.gnu.org/pipermail/libstdc++-cvs/2023q1/039390.html that there might be?

phoebewang replied:
Yes, we may need to add it to compiler-rt too.

shiltian replied:
> I'm also interested in test with -mattr=+avx512bf16,+avx512vl, because the VCVTNEPS2BF16 doesn't raise exception. We should lower it to libcall too.

I'm not familiar with X86 backend. That can be done in a follow-up patch by someone with that expertise.

shiltian commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -393,7 +393,7 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
 
   for (auto Op : {ISD::FP16_TO_FP, ISD::STRICT_FP16_TO_FP, ISD::FP_TO_FP16,
-                  ISD::STRICT_FP_TO_FP16}) {
+                  ISD::STRICT_FP_TO_FP16, ISD::STRICT_FP_TO_BF16}) {
```
@phoebewang Do you mean here, we need to drop `Custom` for `Subtarget.hasF16C()` if `Op` is `ISD::STRICT_FP_TO_BF16`? If yes, I will take it out of the loop.

phoebewang replied:
Yes.

shiltian replied:
I don't know how to implement the `Custom` path since I'm not familiar with what instructions need to be used. I leave a FIXME there.

shiltian commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -437,6 +437,13 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
   }
 
+  // FIXME: If a target has F16C, it needs to be Custom.
+  setOperationAction(ISD::STRICT_FP_TO_BF16, MVT::f32, Expand);
```
Someone with X86 backend expertise needs to change this to `Custom` and implement `case ISD::STRICT_BF16_TO_FP:  return LowerBF16_TO_FP(Op, DAG);` in `LowerOperation` below.

phoebewang replied:
The comment is not correct. We don't need to custom for F16C. We cannot do custom at all because we don't have BF16 truncation instrcutions that support exception.
That's why I suggest you add a test for AVX512BF16, e.g.,
```
; RUN: llc < %s -mtriple=x86_64-linux-gnu -mattr=+avx512bf16,+avx512vl | FileCheck %s --check-prefix=X64
```

phoebewang replied:
I think we don't need to exclude `f80` and `f128`. We can use a loop for them.
```
for (auto VT : {MVT::f32, MVT::f64, MVT::f80, MVT::f128}) {
  setOperationAction(ISD::STRICT_FP_TO_BF16, VT, Expand);
  setOperationAction(ISD::STRICT_BF16_TO_FP, VT, Expand);
}
```

shiltian commented on compiler-rt/lib/builtins/fp_extend.h:
```diff
@@ -81,6 +81,13 @@ static inline int src_rep_t_clz_impl(src_rep_t a) {
 
 #define src_rep_t_clz src_rep_t_clz_impl
 
+#elif defined SRC_BFLOAT
```
The changes in `compiler-rt` were from https://reviews.llvm.org/D151436.

phoebewang replied:
I doubt whether the implementation handles exception or not. But I don't have any suggestion to it.

phoebewang commented on llvm/test/CodeGen/X86/bfloat-constrained.ll:
```diff
@@ -0,0 +1,180 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -mtriple=i686-linux-gnu | FileCheck %s --check-prefix=X32
```
Please X86 instead of X32. X32 is used for x32 mode.

shiltian replied:
Borrowed from `llvm/test/CodeGen/X86/half-constrained.ll` as well. 😄

phoebewang replied:
Solved by https://github.com/llvm/llvm-project/commit/ccc48d45b832def14c8bc1849cf64c805892368d :)

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -424,7 +424,7 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
 
   for (auto Op : {ISD::FP16_TO_FP, ISD::STRICT_FP16_TO_FP, ISD::FP_TO_FP16,
-                  ISD::STRICT_FP_TO_FP16}) {
+                  ISD::STRICT_FP_TO_FP16, ISD::STRICT_FP_TO_BF16}) {
```
This is not needed.

shiltian commented on llvm/test/CodeGen/X86/bfloat-constrained.ll:
```diff
@@ -0,0 +1,180 @@
+; X32-NEXT:    movzwl a, %eax
+; X32-NEXT:    movl %eax, (%esp)
+; X32-NEXT:    calll __extendbfsf2
+; X32-NEXT:    addl $12, %esp
```
I checked the DAG. The extend node was indeed inserted, but I'm not sure why there is nothing emitted. Looking at `llvm/test/CodeGen/X86/half-constrained.ll`, it seems like for some reason on X86 target it just looks like that.

phoebewang replied:
Oh, I see. It is using x87 registers. IIRC, the data is preserved in high precision in register. So we save an extra conversion.

phoebewang commented on llvm/test/CodeGen/X86/bfloat-constrained.ll:
```diff
@@ -0,0 +1,245 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -mtriple=i686-linux-gnu | FileCheck %s --check-prefix=X86
+; RUN: llc < %s -mtriple=x86_64-linux-gnu | FileCheck %s --check-prefix=X64
+; RUN: llc < %s -mtriple=x86_64-linux-gnu -mattr=+avx512bf16,+avx512vl | FileCheck %s --check-prefix=X64-AVX512
```
I see the only different between them are vaddss vs. addss. I think we can add a avx2 to the first one:
```
; RUN: llc < %s -mtriple=x86_64-linux-gnu -mattr=+avx2 | FileCheck %s --check-prefix=X64
; RUN: llc < %s -mtriple=x86_64-linux-gnu -mattr=+avx512bf16,+avx512vl | FileCheck %s --check-prefix=X64-AVX512
```
or
```
; RUN: llc < %s -mtriple=x86_64-linux-gnu | FileCheck %s --check-prefixes=X64,X64-SSE
; RUN: llc < %s -mtriple=x86_64-linux-gnu -mattr=+avx512bf16,+avx512vl | FileCheck %s --check-prefixes=X64,X64-AVX512
```

phoebewang reviewed (approved):
LGTM with one nit, thanks!

shiltian committed:
[SelectionDAG] Add `STRICT_BF16_TO_FP` and `STRICT_FP_TO_BF16`

This patch adds the support for `STRICT_BF16_TO_FP` and `STRICT_FP_TO_BF16`.

Fix #78540.
